lib/handler/traffic.c

```diff
@@ -0,0 +1,144 @@
+#include <stdlib.h>
+#include "h2o.h"
+
+#ifndef HUNDRED_MS
+#define HUNDRED_MS 100
+#endif
+
+#ifndef ONE_SECOND
+#define ONE_SECOND 1000
+#endif
+
+typedef H2O_VECTOR(h2o_iovec_t) iovec_vector_t;
+
+typedef struct st_traffic_shaper_t {
+    h2o_ostream_t super;
+    h2o_timeout_entry_t timeout_entry;
+    size_t tokens;
+    size_t token_inc;
+    h2o_context_t *ctx;
+    h2o_req_t *req;
+    struct {
+        iovec_vector_t bufs;
+        size_t inbufcnt;
+        int is_final;
+    } state;
+} traffic_shaper_t;
+
+static int real_send(traffic_shaper_t *self) {
+    /* a really simple token bucket implementation */
+    assert(self->tokens);
+    size_t i, token_consume;
+
+    token_consume = 0;
+
+    for (i = 0; i < self->state.inbufcnt; i++) {
+        if (self->state.bufs.entries[i].len > self->tokens - token_consume) {
+            return 0;
+        }
+        token_consume += self->state.bufs.entries[i].len;
+    }
+
+    self->tokens -= token_consume;
+
+    h2o_ostream_send_next(&self->super, self->req, self->state.bufs.entries,
+                          self->state.inbufcnt, self->state.is_final);
+    return self->state.is_final;
+}
+
+static inline void expand_buf(h2o_mem_pool_t *pool, iovec_vector_t *bufs, size_t count) {
+    assert(bufs->size < count);
+    h2o_vector_reserve(pool, bufs, count);
+    bufs->size = count;
+}
+
+static void add_token(h2o_timeout_entry_t *entry) {
+    traffic_shaper_t *shaper = H2O_STRUCT_FROM_MEMBER(traffic_shaper_t, timeout_entry, entry);
+    shaper->tokens += shaper->token_inc;
```

Comment:
We might need to take care of overflow here, especially in case `tokens` is a 32-bit value.
For example in case the bitrate is 10MB/sec and `tokens` is a 32-bit value, and if the upstream is stalling, `tokens` would overflow in 409 seconds.

One easy fix to the issue would be to change `tokens` to a 64-bit type.